File: fastxc/sac2spec_pyv/arguproc.py
import argparse
import shlex


def parse_arguments(command_line=None):
    parser = argparse.ArgumentParser(
        description="Process SAC files to generate spectrum files with optional GPU acceleration.",
        epilog="Last updated by wangjx@20240710",
    )
    parser.add_argument(
        "-I",
        dest="sac_lst",
        required=True,
        help="Input list of sac files of multi/single components.",
    )
    parser.add_argument(
        "-O",
        dest="spec_lst",
        required=True,
        help="Output spec list of segment spectrum files.",
    )

    # Optional arguments
    parser.add_argument(
        "-C", dest="num_ch", type=int, help="Number of channels to process."
    )
    parser.add_argument(
        "-B", dest="filter_file", help="File containing Butterworth filter parameters."
    )
    parser.add_argument(
        "-L",
        dest="seglen",
        type=float,
        required=True,
        help="Length of segment window in seconds, usually 7200s (2 hours).",
    )
    parser.add_argument("-G", dest="gpu_id", type=int, help="Index of GPU device.")
    parser.add_argument(
        "-F",
        dest="freq_band",
        type=parse_frequency_band,
        help="Frequency bands for spectral whitening in Hz, using the format f_low/f_high.",
    )
    parser.add_argument(
        "-W",
        dest="whiten_type",
        type=int,
        choices=[0, 1, 2, 3],
        help="Whiten type. 0: no whitening; 1: pre-time domain normalization; 2: post-time domain normalization; 3: both.",
    )
    parser.add_argument(
        "-N",
        dest="normalize_type",
        type=int,
        choices=[0, 1, 2],
        help="Normalization type. 0: no normalization; 1: runabs; 2: one-bit.",
    )
    parser.add_argument(
        "-Q",
        dest="skip_steps",
        type=parse_skip_steps,
        help="Skip segment steps. Use -1 to indicate the end of input sequence.",
    )
    parser.add_argument(
        "-T",
        dest="thread_num",
        type=int,
        default=1,
        help="GPU thread number. Default is 1.",
    )
    parser.add_argument("-U", dest="gpu_num", type=int, help="Number of GPUs to use.")
    
    if command_line:
        # 使用 shlex.split 来智能拆分命令行参数字符串
        args = parser.parse_args(shlex.split(command_line))
    else:
        # 没有提供 command_line 时，从命令行读取参数
        args = parser.parse_args()
        
    return args


def parse_frequency_band(value):
    try:
        freq_low, freq_high = map(float, value.split("/"))
        if freq_low >= freq_high:
            raise argparse.ArgumentTypeError("Invalid frequency band range")
        return freq_low, freq_high
    except ValueError:
        raise argparse.ArgumentTypeError("Invalid frequency band format")


def parse_skip_steps(value):
    try:
        steps = []
        for val in value.split("/"):
            num = int(val)
            if num == -1:
                break  # Stop parsing if -1 is found
            steps.append(num)
        return steps
    except ValueError:
        raise argparse.ArgumentTypeError("Invalid skip step format")


if __name__ == "__main__":
    args = parse_arguments()
    print(
        args
    )  # Just a placeholder for demonstration. Replace with actual function calls.

File: src/sac2spec_cos/arguproc.c
#include "arguproc.h"

void ArgumentProcess(int argc, char **argv, ARGUTYPE *pargument)
{
    if (argc < 2)
    {
        usage();
        exit(1);
    }
    int opt;

    // Set default values for optional arguments
    pargument->whitenType = 0;
    pargument->normalizeType = 0;
    pargument->skip_step_count = 0;
    pargument->thread_num = 1;

    // Parse command line arguments using getopt
    while ((opt = getopt(argc, argv, "I:O:C:B:L:G:F:W:N:Q:T:U:")) != -1)
    {
        switch (opt)
        {
        case 'I':
            pargument->sac_lst = optarg;
            break;
        case 'O':
            pargument->spec_lst = optarg;
            break;
        case 'C':
            pargument->num_ch = atoi(optarg);
            break;
        case 'B':
            pargument->filter_file = optarg;
            break;
        case 'L':
            pargument->seglen = atof(optarg);
            break;
        case 'G':
            pargument->gpu_id = atoi(optarg);
            break;
        case 'F':
        {
            float freq_low, freq_high;
            if (sscanf(optarg, "%f/%f", &freq_low, &freq_high) != 2)
            {
                fprintf(stderr, "Error: Invalid frequency band format\n");
                exit(1);
            }

            // ensure freq_low_limit < freq_high_limit
            if (freq_low >= freq_high)
            {
                fprintf(stderr, "Error: Invalid frequency band range\n");
                exit(1);
            }
            pargument->freq_low = freq_low;
            pargument->freq_high = freq_high;
            break;
        }
        case 'W':
            pargument->whitenType = atoi(optarg);
            break;
        case 'N':
            pargument->normalizeType = atoi(optarg);
            break;
        case 'Q':
        {
            char *token = strtok(optarg, "/");
            pargument->skip_step_count = 0;
            while (token != NULL && pargument->skip_step_count + 1 < MAX_SKIP_STEPS_SIZE)
            {
                int val = atoi(token);
                if (val == -1)
                {
                    break; // stop parsing when find -1
                }
                pargument->skip_steps[pargument->skip_step_count++] = val;
                token = strtok(NULL, "/");
            }
            break;
        }
        case 'U':
            pargument->gpu_num = atoi(optarg);
        case 'T':
            pargument->thread_num = atoi(optarg);
            break;
        default: // '?' or ':' for unrecognized options or missing option arguments
            usage();
            exit(-1);
        }
    }
}

/* display usage */
void usage()
{
    fprintf(stderr, "Usage:\n"
                    "sac2spec   -I filelist -O outputlist -L winlength \n"
                    "           -C channel_numbers -G gpu_no \n"
                    "           -W whitenType -N normalizeType -F f1/f2 -B filter_coefficients_file \n"
                    "           -U gpu_para_numbers -T num_of_cpu -Q skip_step\n"
                    " Options:\n"
                    "   -I   input list of sac files of multi/single components.\n"
                    "   -O   output spec list of segment spectrum files. Corresponding to -I option\n"
                    "   -L   length of segment window in seconds.\n"
                    "   -C   number of channels in the input sac files.\n"
                    "   -G   Index of Gpu device \n"
                    "   -F   FREQ_BANDS Frequency bands for spectral whitenning (in Hz) \n"
                    "        using the format f_low_limit/f_high_limit \n"
                    "   -W   Whiten type.\n"
                    "		   0: donot do any whitenning\n"
                    "		   1: do whitening BEFORE time domain normalization\n"
                    "		   2: do whitening AFTER time domain normalization\n"
                    "          3: do BEFORE and AFTER time domain normalization\n"
                    "   -N    Normalization type\n"
                    "          0: no normalization.\n"
                    "          1: runabs normalization.\n"
                    "          2: one-bit normalization.\n"
                    "   -B  Butterworth coefficient file. The filter file should be generated \n"
                    "   -Q   skip segment step. Default is 1, which means we do not skip any step\n"
                    "   -T  [optional >=0] CPU thread number. Default is 1\n"
                    "   -U  [optional >=0] GPU thread number. Default is 1\n"
                    "   Last updated by wangjx@20241127\n");
}